client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+The cache should utilize an HTTP Request/Response library (e.g., in Python, urllib3) to establish a connection with the DB server and read and write values through its course of use.  We describe in the HTTP Message Formatting section the expected format for DB requests and responses.
+
+[Here are examples](https://urllib3.readthedocs.org/en/latest/) of using the urllib3 library in Python to open and use a connection.
+
+### Sorted Key-Value Map of Cache Entries
+
+The cache is a ("table:key" map) of (sorted value maps), which we explain below.
+
+The keys of the "table:key" map are composed of the concatenation of the table identifier string and the key string; each "table:key" maps to a (sorted value map).  (Note that the cache may store entries from multiple tables.)  
+
+The cache entries themselves are stored in (sorted value maps).  The keys of the (sorted value maps) are derived from the read timestamps of each entry.  Since we want the entries sorted from most-recent to oldest read timestamps, the actual keys are longs with the value 
+
+key = (LONG\_MAX - read\_timestamp)
+
+For example, in Python, the outer "table:key" map can be implemented as a standard Python dictionary, and we just use string concatenation (a+b) to form the keys.  For the inner sorted value maps, we can use a SortedDict implementation based on [skiplists](https://github.com/sepeth/python-skiplist).
+
+* **Question (1):** We said Tuesday we just need a single-level map, but why don't we need at least the two levels I describe here?
+
+## Cache Public Method Descriptions
+
+### read
+    read(
+        table: string, 
+        key: string, 
+        read_timestamp=currentTime, 
+        max_age=None, 
+        no_cache=False): value (JSON)  
+
+The read method sends a read request to the DB for the given table and key, along with the specified cache parameters, if any.  It returns the value object if the DB provides a value; otherwise, it returns None.
+
+In the case read returns None (value not in database), we still update our cache with the value entry None for the specified key at the read\_timestamp.  The None value may represent an update to the key.
+
+Note that the max\_age, and no\_cache parameters can be specified in at least three separate locations: (1) in the Cache constructor, (2) in the Cache read method, and (3) in the Transaction read method.  The cache read method determines the most restrictive set of parameters based on these three inputs and sends this most restrictive parameter set to the DB.  With respect to the read\_timestamp parameter, the DB server's response to every query includes the read\_timestamp it utilized to fulfill the request; we use the read\_timestamp the server returns to us to update the cache.
+
+Here is the method for determining the most restrictive parameter set:
+
+* **read\_timestamp**: The most restritive read\_timestamp is the minimum of the given read\_timestamp values (i.e., min(cache, transaction) read\_timestamps)
```

Comment:
The read time doesn't fit into this part of the discussion. There's no sense of "most restrictive read." The read time is an arg to the tx constructor, and tx read passes it through to cache.read. The user specifies the read-time only in the tx constructor, not in it's read method and not in the cache-constructor.